smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
         float
             Value of the derivative of Trigonometric polynomial.
 
+        n : int, optional
+            order of derivative. Default is 1.
+
         Raises
         ------
         ValueError
             If input is outside the domain `[0, 2\pi]`
         """
         x = numpy.asarray(x)
-        return self.frequency * 1j * numpy.exp(x * self.frequency * 1j)
+        return numpy.exp(x * self.frequency * 1j) * (self.frequency * 1j) ** n
+
+
+class BasisFunctionSet(collections.abc.Sequence):
+    """Set of basis functions and sample points."""
+
+    @abc.abstractmethod
+    def __init__(self):
+        pass
+
+    @property
+    def basis_functions(self):
+        """list: Basis functions."""
+        return self._basis_functions
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `basis_functions`"""
+        return self.basis_functions[0].domain
 
+    def __len__(self):
+        return len(self.basis_functions)
 
-class BasisFunctionSet:
-    """Set of basis functions and sample points.
+    def __getitem__(self, key):
+        return self.basis_functions[key]
+
+    def _scale_to_domain(self, points, old_domain):
+        points = self.domain[0] + (self.domain[1] - self.domain[0]) * (
+            (points - old_domain[0]) / (old_domain[1] - old_domain[0])
+        )
+        return numpy.clip(points, self.domain[0], self.domain[1])
```

Comment:
If you know all the domains are the same, then a pass through is fine. You will want to add a line that checks if the basis functions are not empty first though. If that's a common task, you can put it in a private method.
```suggestion
    def scale_to_domain(self, points, old_domain):
        return self.basis_functions[0].scale_to_domain(points, old_domain)
```